Fix usage of seekAsync in MessageImpl.hasMessageAvailable and flaky-test

315157973 opened the pull request:
Fixes #10152

### Motivation

##### 1
When the connection is reset, message received by reader may be duplicated, so I remove the judgment of whether the key is unique
Repeated scene:

start to read -> 1255:0 threadId:68
after read,update read position -> 1255:1 threadId:68
start to read -> 1255:1 threadId:68

[pulsar-external-listener-39-1] WARN  org.apache.pulsar.client.impl.ConsumerImpl - [persistent://my-property/my-ns/my-reader-topic-with-batching-inclusivebdcf4a3e-6c2e-4d36-b362-52c066d14da0-partition-2] [d975f7e2-7b8f-467a-a56b-b3c1e96ec75f] Could not get connection while getLastMessageId -- Will try again in 190 ms

[pulsar-external-listener-39-1] INFO  org.apache.pulsar.client.impl.ConsumerImpl - [persistent://my-property/my-ns/my-reader-topic-with-batching-inclusivebdcf4a3e-6c2e-4d36-b362-52c066d14da0-partition-2][d975f7e2-7b8f-467a-a56b-b3c1e96ec75f] Get topic last message Id

[pulsar-client-io-38-1] INFO  org.apache.pulsar.client.impl.ConsumerImpl - [persistent://my-property/my-ns/
my-reader-topic-with-batching-inclusivebdcf4a3e-6c2e-4d36-b362-52c066d14da0-partition-2][d975f7e2-7b8f-467a-a56b-b3c1e96ec75f] Successfully getLastMessageId 1255:0

[pulsar-client-io-38-1] INFO  org.apache.pulsar.client.impl.ConsumerImpl - [persistent://my-property/my-ns/my-reader-topic-with-batching-inclusivebdcf4a3e-6c2e-4d36-b362-52c066d14da0-partition-2][d975f7e2-7b8f-467a-a56b-b3c1e96ec75f] **Seek subscription to the message 1255:0:2:4**

##### 2
When I fixed bug 1, I found bug 2
normal situation:
1) ConsumerImpl.hasMessageAvailableAsync triggers seek, and then seek triggers reconnection
2) Reconnection triggers connectionOpened, at this time startMessageId = clearReceiverQueue() = seekMessageId
3) ConsumerImpl reconnects. BrokerlastDequeuedMessageId=earliest, startMessageId=seekMessageId
4) hasMessageAvailableAsync return true
5) The broker returns a message
6) Consumer receive

abnormal situation:
1) ConsumerImpl.hasMessageAvailableAsync triggers seek, and then seek triggers reconnection
2) Reconnection triggers connectionOpened, at this time startMessageId = clearReceiverQueue() = seekMessageId
3) ConsumerImpl reconnects. BrokerlastDequeuedMessageId=earliest, startMessageId=seekMessageId
4) hasMessageAvailableAsync return true
5) The broker returns a message
6) Due to other reasons, the connection is disconnected again, and step 2, 3, and 4 are triggered again.
When 2 is triggered, the message in incomeQueue is cleaned, and startMessageId is no longer equal to latest, so seek will not be triggered anymore, resulting in the cleaned message cannot be read

The only time to trigger seek now is:
lastDequeuedMessageId == MessageId.earliest
and
startMessageId.equals(MessageId.latest)
and
resetIncludeHead == true

### Modifications
1）remove ·`assertTrue(set.remove(key))` in uniteTest, because message may be duplicated
2）add seek in hasMessageAvailableAsync

eolivelli commented:
What about changing the Title and explicitly point out that we are modifying the behaviour of "hasMessageAvailableAsync" ?

The title reads about fixing a flaky test but we are touching  ConsumerImpl

315157973 commented:
> What about changing the Title and explicitly point out that we are modifying the behaviour of "hasMessageAvailableAsync" ?
> 
> The title reads about fixing a flaky test but we are touching ConsumerImpl

OK, do you have any good suggestions ?

eolivelli commented:
What about "Fix usage of seekAsync  in MessageImpl.hasMessageAvailable" ?

315157973 commented:
> What about "Fix usage of seekAsync in MessageImpl.hasMessageAvailable" ?

Em...But I fixed 2 bugs in total, not only `hasMessageAvailable` but also including flaky-test. If I submit flaky-test separately, there will be other problems.

merlimat commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:
```diff
@@ -1863,8 +1863,17 @@ public boolean hasMessageAvailable() throws PulsarClientException {
+                // 3.Broker has pushed messages to ReceiverQueue, but messages were cleaned due to connection reset
+                seekAsync(startMessageId).whenComplete((ignore, e) -> {
+                    if (e != null) {
+                        booleanFuture.complete(false);
```
Couldn't we have the case here where the seekAsync fails and we return `false` for `hasMessageAvailable()`, when in reality there may be some message?

codelipenghui reviewed (commented):
Seems the root cause is the reader uses a incorrect start message ID to subscribe to a topic during the reconnection. I think the right way to fix this problem is fix the start message ID that the reader used for subscribe to the topic and the topic always apply the reader's start message ID since the reader always uses the exclusive non-durable subscription. And the case of the reader never consume any messages due to the reconnection and users uses the latest position as the initial start message ID, this might introduce uncertain behavior during the reconnection. My suggestion is if uses the latest positioin as the nitial start message ID for a reader, we can get the last message ID from the broker first and then use this message ID as the start message ID to subscribe to the topic, so that as long as the consumer does not crashes, it will always uses the certain message ID to subscribe to the topic when it never receives any messages.

315157973 commented:
Indeed, if we put the seek in this place, at this time, there is a high probability that the connection is not ready, causing the seek to fail, and eventually this bug still exists. It would be better to put it in connectionOpen.
We can use subscription to set the initial position.

315157973 commented:
I tried to handle this situation in connectionOpen, but failed:
1 The first time call hasMessageAvailable, it will get the last msg in broker, and then trigger seek lastMsgIdInBroker
2 seek will trigger disconnect and reconnect, clearReceiverQueue() will be called when connectionOpen, and then startMessageId will be assigned to seekId (lastMsgIdInBroker)
3 Then connection is disconnected again(For other unknown reasons), and the startMessageId subscribed at this time is lastMsgIdInBroker, and the broker will not return this message.

Because the sub command does not have an inclusive option, it can only be read from the next item of startMessageId. And I can't set entryId or ledger -1, otherwise the client may read forward after multiple reconnections. Unless I change the proto, add the inclusive option.

@codelipenghui

315157973 commented:
Record the process:
1) I modified the proto so that the subscription supports the inclusive option. I found that `hasMessageAvailableAsync` and `messageProcessed` are not in the same thread, which has thread safety issues with seek. Time line:
```
messageProcessed 1131:0:0:4 topic 0 
Successfully getLastMessageId 1131:0
Seek subscription to the message 1131:0:0:4
connectionOpened 1131:0:0:4 topic 0
hasMoreMessages 1131:0:0:4 lastMessageIdInBroker 1131:0:0:4 lastDequeuedMessageId-1:-1:-1 topic 0
hasMoreMessages 1131:0:0:4 lastMessageIdInBroker 1131:0:0:4 lastDequeuedMessageId-1:-1:-1 topic 0
message received [1132:0:1:4, 1131:0:0:4, 1133:0:2:4, 1133:0:2:4]
```

2) Then I implemented the first seek in ConnectionOpen, and found that when there is no subscription, the lastMessageIdInBroker cannot be obtained.

Finally, I can only temporarily use seek to solve this problem. If anyone wants to use subscription to solve this problem, they can open another issue.

315157973 commented:
@eolivelli @merlimat @codelipenghui  PTAL

315157973 commented on pulsar-client/src/main/java/org/apache/pulsar/client/util/RetryUtil.java:
```diff
@@ -0,0 +1,65 @@
+        if (backoff.getInitial() <= 0) {
+            throw new IllegalArgumentException("Illegal initial time");
+        }
+        scheduledExecutorService.execute(() ->
```
The first time is to run directly, only when an exception occurs, it will wait before retrying

315157973 commented on pulsar-client/src/main/java/org/apache/pulsar/client/util/RetryUtil.java:
```diff
@@ -0,0 +1,65 @@
+            if (isMandatoryStop) {
+                callback.completeExceptionally(e);
+            } else {
+                log.info("RetryUtil will retry in {} ms", next);
```
Good idea, I will fix it

eolivelli reviewed (approved):
LGTM

I believe this fix does not fix 2.7.2, it is better to keep it only on master branch and give time to settle

eolivelli commented:
@codelipenghui can you please take a final look ?

315157973 commented:
This PR can also fix these problems:
1) hasMessageAvailable returns true, but the reader cannot read any message
2) After the broker restarts, the reader will not continue to receive messages
